### ISSUE TITLE
text and write function are buggy

### ISSUE THREAD

**Issue by dobodu** [dobodu/RM690B0_Micropython_QSPI#1]
The TEXT (fixed font) or WRITE (variable width) do present bugs.

See picture for more

**Comment by dobodu** [comment]
![20240707_154333~2.JPG](https://github.com/dobodu/RM69690_Micropython_QSPI/assets/8106589/59c92bab-6c57-4ca6-b9d5-5d0e9193acbb)

**Comment by dobodu** [comment]
tft.text(font, b'Hello!', x, y , fontcolor, backcolor)

**Bug appears when x is odd,** when x is even the string shows correctly

**Comment by nspsck** [comment]
Hi, 

TL;DR: Sorry, I can not help you.

The reason:

I think the issue lies in [here](https://github.com/dobodu/RM69690_Micropython_QSPI/blob/da49c71f621a1ce1b8095a5ff392569071a23a0d/rm690b0/rm690b0.c#L374-L394). The `set_area()` function sets the area of the memory region of the display controller to be updated when inputs come in. 
```C
    uint8_t bufx[4] = {
        ((x0 >> 8) & 0xFF),
        (x0 & 0xFF),
        ((x1 >> 8) & 0xFF),
        (x1 & 0xFF)};
    uint8_t bufy[4] = {
        ((y0 >> 8) & 0xFF),
        (y0 & 0xFF),
        ((y1 >> 8) & 0xFF),
        (y1 & 0xFF)};
```
These two arrays specify the area and the offset value `0xFF` depends on the display controller's specification. This is not necessarily a universal value, hence you might want to check the user manual. (But that is probably not going to be very helpful.)

The user manual states:
![image](https://github.com/dobodu/RM69690_Micropython_QSPI/assets/86347287/5ae9be2e-36f5-4533-bd56-980aba314de5)

`2A00h` to `2A01h` controls the `x0`, they claims only `SC[9:0]` are cared, the rest is no-care. But that does not make any sense if this is accurate and the display's memory is point-to-point to the display pixels since you only have 10 bits and all you can control with it are
512 pixels at most. Hence the `SC[10]` has to be somehow used, or some magic is used in `RAMWR`, which is very unlikely to happen. Because this is a limitation for both row and column, if `SC[10]` wasn't used, you will never have precise control over the 513th pixel. 

What I am trying to say is: that I have no idea what is going on...


Tho I have a suspicion. This is just a suspicion.

There are 2 restrictions for `CASET` and `RASET`:
> (1) SC[9:0] always must be equal to or less than EC[9:0].
(2) The SC[9:0] and EC[9:0]-SC[9:0]+1 must can be divisible by 2.

In the `text()` function you see:
```C
                uint16_t x1 = x0 + width - 1;
                if (x1 < self->width) {
                    set_area(self, x0, y0, x1, y0 + height - 1);
                    write_color(self, (uint8_t *)self->frame_buffer, buf_size);
                }
```
`width` is the font's width and it is always even. When `x0` is odd, `x1` becomes even, so `x1 - x0 + 1` is even. When `x0` is even, `x1` becomes odd, so `x1 - x0 + 1` is even. So we can rule out the second one. Maybe due to some memory magic, the first restriction is not met. However, I cannot see how this could happen.

**Comment by dobodu** [comment]
> These two arrays specify the area and the offset value `0xFF` depends on the display controller's specification. This is not necessarily a universal value, hence you might want to check the user manual. (But that is probably not going to be very helpful.)

I've been checking and FF is the correct value for RM690B0

> There are 2 restrictions for `CASET` and `RASET`:
> 
> > (1) SC[9:0] always must be equal to or less than EC[9:0].
> > (2) The SC[9:0] and EC[9:0]-SC[9:0]+1 must can be divisible by 2.
> 

I've seen that, but why should RM690B0 an RM67162 act different. Both manuals requires the same. And I never had any problem writing to any location within RM67162 library.

I'm digging a bit in the code to understand

**Comment by dobodu** [comment]
A thread explaining the odd/even problem

https://github.com/Xinyuan-LilyGO/T-Display-S3-AMOLED/issues/2

**Comment by nspsck** [comment]
Ugh, thank you for providing the thread! I finnaly read through that thread, seems like nothing new was mentioned to this issue. The driver uses only 16-bits color565 mode and 8bits also 24bits are not implemented. Tho, it let you to choose 8bits or 24bits, but the color will be wrong. TBH, I only checked the `text()` function, and the logics should work.

Also, the 2 rules about odd/even restrictions should be met as I have explained before:
> In the text() function you see:

                uint16_t x1 = x0 + width - 1;
                if (x1 < self->width) {
                    set_area(self, x0, y0, x1, y0 + height - 1);
                    write_color(self, (uint8_t *)self->frame_buffer, buf_size);
                }
> width is the font's width and it is always even. When x0 is odd, x1 becomes even, so x1 - x0 + 1 is even. When x0 is even, x1 becomes odd, so x1 - x0 + 1 is even. So we can rule out the second one. Maybe due to some memory magic, the first restriction is not met. However, I cannot see how this could happen.

Tho, his explainations makes sense and suit the situation in the picture you provided:
> Failure to 'round off' your update regions to meet these criteria will result in the kind of intermittent 45 degree skewing that looks like tearing but is in fact a restriction of the display hardware. Your screen shots look to me like this is what was happening. A full screen update avoids the issue as 536x240 meets these criteria.

Then I double checked the  `write()` function, which, even though not the exact same expression, but the drawing part is essentially the same thing like the `text()` function. Which adhere to the 2 rules. It is really weird...

**Comment by dobodu** [comment]
> Then I double checked the `write()` function, which, even though not the exact same expression, but the drawing part is essentially the same thing like the `text()` function. Which adhere to the 2 rules. It is really weird...

I implemented a Odd/Even correction in both fill_rect and text function.
It does correct the very first bug we had (so no more need on framebuffer=false).
I'll implement the write function in the very next days.

I think it does also fix the issue within the RM67162. As a conclusion, there wasn't any bug in our codes, but only a misunderstanding in the way the RM67162 and RM690B0 handles memory copy.

**Comment by dobodu** [comment]
I've implemented the correction I made to text() function in the write() function but I'm too lazy tonight to test the result.
Files compiled straight (only 2 warning)... that's usually not a good news ;-)

**Comment by nspsck** [comment]
As long as the firmware is built, you can sleep in peace. 😆
